# Cleaning: K-Nearest Neighbors & Credit Card Fraud

## **1.** Imports & Settings

In [ ]:
import pandas as pd
import numpy as np

# scalers used in section 5.2.
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import PowerTransformer


pd.set_option('display.max_rows', 200)

---

## **2.** Load Data
Here we load the data from Google Drive, but the data can also be accessed from a relative path for use with Jupyter Notebooks/Lab.


### **2.1.** Load Data w/ Jupyter Notebooks/Lab
```
# run this code if using Jupyter Notebooks/Lab
df = pd.read_csv('./data/creditcard.csv')
```

### **2.2.** Access & Load Data w/ Google Drive

In [ ]:
# run this cell if using Google Colab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
path = '/content/drive/MyDrive/Colab Notebooks/KNN_creditCardFraud/data/creditcard.csv'
df = pd.read_csv(path)

---

## **3.** Detail & View Data
Below we look into details that inform the cleaning, scaling, and subsetting that we may need to perform. This includes looking for unusual or null values and checking the distribution of our classifications.

### **3.1.** Head & Info
The Kaggle description of our dataset indicates that "Time" and "Amount" are the only two columns that did not undergo PCA transformation. As part of PCA, data is normalized. A quick look at `df.head()` and  `df.info()`confirms that "Time" and "Amount" are on a different scale to our 29 V-n columns. It also reveals: 
*   Our dataset contains 29 V-n Features, plus 'Time', 'Amount', and 'Class'
*   There are 0 Null values
*   There are two classes
    * 0 Represents "Not Fraud"
    * 1 Represents "Fraud"
* All Features are floats and Class is int

In [ ]:
df.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

Viewing `df.describe()`, we find that 'Time' and 'Amount' are on different scales to our V-n features. Considering we will experiment with a KNN classifier, these features may be valuable, so we will plan to scale them alongside our other features.

In [ ]:
df.describe().T

count          mean  ...            75%            max
Time    284807.0  9.481386e+04  ...  139320.500000  172792.000000
V1      284807.0  3.919560e-15  ...       1.315642       2.454930
V2      284807.0  5.688174e-16  ...       0.803724      22.057729
V3      284807.0 -8.769071e-15  ...       1.027196       9.382558
V4      284807.0  2.782312e-15  ...       0.743341      16.875344
V5      284807.0 -1.552563e-15  ...       0.611926      34.801666
V6      284807.0  2.010663e-15  ...       0.398565      73.301626
V7      284807.0 -1.694249e-15  ...       0.570436     120.589494
V8      284807.0 -1.927028e-16  ...       0.327346      20.007208
V9      284807.0 -3.137024e-15  ...       0.597139      15.594995
V10     284807.0  1.768627e-15  ...       0.453923      23.745136
V11     284807.0  9.170318e-16  ...       0.739593      12.018913
V12     284807.0 -1.810658e-15  ...       0.618238       7.848392
V13     284807.0  1.693438e-15  ...       0.662505       7.126883
V14     284807.0  1.479045e-15  ...       0.493150      10.526766
V15     284807.0  3.482336e-15  ...       0.648821       8.877742
V16     284807.0  1.392007e-15  ...       0.523296      17.315112
V17     284807.0 -7.528491e-16  ...       0.399675       9.253526
V18     284807.0  4.328772e-16  ...       0.500807       5.041069
V19     284807.0  9.049732e-16  ...       0.458949       5.591971
V20     284807.0  5.085503e-16  ...       0.133041      39.420904
V21     284807.0  1.537294e-16  ...       0.186377      27.202839
V22     284807.0  7.959909e-16  ...       0.528554      10.503090
V23     284807.0  5.367590e-16  ...       0.147642      22.528412
V24     284807.0  4.458112e-15  ...       0.439527       4.584549
V25     284807.0  1.453003e-15  ...       0.350716       7.519589
V26     284807.0  1.699104e-15  ...       0.240952       3.517346
V27     284807.0 -3.660161e-16  ...       0.091045      31.612198
V28     284807.0 -1.206049e-16  ...       0.078280      33.847808
Amount  284807.0  8.834962e+01  ...      77.165000   25691.160000
Class   284807.0  1.727486e-03  ...       0.000000       1.000000

[31 rows x 8 columns]

### **3.2.** Classification Distrobution
We can plot a histogram to see the distribution of our two classifications (fraud & not-fraud), but because we are working with a binary classification, printing the percentages works just as well:

In [ ]:
# quantify the balance of classifcations
no_fraud_percent = round((df['Class'].value_counts()[0] / len(df)) * 100, 2)
fraud_percent = round((df['Class'].value_counts()[1] / len(df)) * 100, 2)

print(f'No Fraud: {no_fraud_percent} percent of dataset')
print(f'Fraud: {fraud_percent} percent of dataset')

No Fraud: 99.83 percent of dataset
Fraud: 0.17 percent of dataset


We find our classifications to be heavily skewed in favor of legitimate transactions, so we will also plan to balance our data to equal observations of fraud and not-fraud cases.
* 99.83% of the data represented by "No Fraud"
* .17% classified as "Fraud"

## **4.** Adjust Dataset
We will adjust our dataset according to the findings above—particularly in respect to the unbalanced classifications and additionally in scaling our features.

### **4.1.** Create Subset
We create a subset of our original dataset, balanced according to "Class", so that we have an even number of fraud and not-fraud samples. Using our original DataFrame, with such an imbalance in favor of not-fraud observations, would potentailly create a bias in our model.
<br />
<br />
Balancing our data will also allow us to create an informative correlation matrix during our EDA. If we were to use the original DataFrame, the correlation matrix would be similarily influenced by the large imbalance between classes.

In [ ]:
# save n_fraud to pass into our subset process below
n_fraud = df['Class'].value_counts()[1]

n_fraud

492

In [ ]:
# set random_state for consistency in subsequent experiments
df_sample = df.sample(frac=1, random_state=4)

# sample an equal number of fraud and non-fraud observations
fraud_df = df_sample.loc[df_sample['Class'] == 1]
non_fraud_df = df_sample.loc[df_sample['Class'] == 0][:n_fraud]

distributed_df = pd.concat([fraud_df, non_fraud_df])

# sample dataframe rows
sample_df = distributed_df.sample(frac=1, random_state=89)

# check distrobution
no_fraud_percent = round((sample_df['Class'].value_counts()[0] / 
                          len(sample_df)) * 100, 2)
fraud_percent = round((sample_df['Class'].value_counts()[1] / 
                       len(sample_df)) * 100, 2)

print(f'No Fraud: {no_fraud_percent} percent of dataset')
print(f'Fraud: {fraud_percent} percent of dataset')
print(f'Length of sample_df: {len(sample_df)}')

No Fraud: 50.0 percent of dataset
Fraud: 50.0 percent of dataset
Length of sample_df: 984


### **4.2.** Scale Subset
Below, we will generate three different CSVs, each scaled by a different method. We will test each to see how different scaling methods influence the final results of our model.

In [ ]:
# set scalers as a dict so that we can loop through them
scale_dict = {'minmax': MinMaxScaler(), 
              'robust': RobustScaler(), 
              'power': PowerTransformer()}

In [ ]:
for key, value in scale_dict.items():
  sub_scale_df = sample_df.copy()

  scaler = value

  # fit and transform the data
  sub_scale_df.iloc[:, :30] = scaler.fit_transform(sub_scale_df.iloc[:, :30])

  # export to csv
  sub_scale_df.to_csv(f'/content/drive/MyDrive/Colab Notebooks/KNN_creditCardFraud/data/{key}_sub.csv', 
                      index=False)
  
  print(f'\n{key}')
  display(sub_scale_df.head())



minmax


Time        V1        V2  ...       V28    Amount  Class
73857   0.322234  0.742574  0.560633  ...  0.233939  0.000390      1
143336  0.496940  0.725715  0.612967  ...  0.172504  0.118972      1
64460   0.298010  0.588966  0.698188  ...  0.239811  0.047035      1
81609   0.343800  0.859250  0.457836  ...  0.275796  0.653643      1
274002  0.966264  0.835847  0.581881  ...  0.204399  0.021168      0

[5 rows x 31 columns]


robust


Time        V1        V2  ...       V28     Amount  Class
73857  -0.312063 -1.375053  0.153832  ...  0.441731  -0.194876      1
143336  0.029278 -1.517784  0.983791  ... -1.535918   2.367795      1
64460  -0.359391 -2.675517  2.335314  ...  0.630739   0.813154      1
81609  -0.269927 -0.387259 -1.476427  ...  1.789126  13.922537      1
274002  0.946243 -0.585397  0.490798  ... -0.509193   0.254143      0

[5 rows x 31 columns]


power


Time        V1        V2  ...       V28    Amount  Class
73857  -0.610685 -1.032163 -0.016067  ...  0.278884 -1.278094      1
143336  0.032848 -1.094698  0.601466  ... -0.918079  1.245622      1
64460  -0.707474 -1.511311  1.523454  ...  0.386018  0.818455      1
81609  -0.526358 -0.461862 -1.568309  ...  1.020385  1.985211      1
274002  1.488746 -0.604199  0.241746  ... -0.279214  0.438772      0

[5 rows x 31 columns]

---

## **5.** Conclusion
 
We've taken an initial look at our data and created three new CSVs to work with when modeling:
1. A MinMax-scaled subset with balanced classifications
2. A Robust-scaled subset with balanced classifications
3. A PowerTransformer-scaled subset with balanced classifications

After adjusting the dataset to account for the imbalance in classifications, we are left with a relatively small amount of data to train a model on: 984 observations. However, the K-Nearest Neighbors model is particularly well-suited for smaller datasets. During our exploration, we hope to find features strongly correlated to our classifications.
<br />
<br />
View our exploratory data analysis in ./EDA-FS_CC-fraud.ipynb